src/c++/libos/lwip/lwip_queue.cc

```diff
@@ -1399,35 +1481,42 @@ dmtr::lwip_queue::parse_packet(struct sockaddr_in &src,
     dst.sin_family = AF_INET;
 
     // segment count
-    sga.sga_numsegs = ntohl(*reinterpret_cast<uint32_t *>(p));
-    p += sizeof(uint32_t);
+    uint32_t numsegs_buf;
+    auto * const numsegs = pktmbuf_struct_read(pkt, offset, numsegs_buf);
+    sga.sga_numsegs = ntohl(*numsegs);
+    offset += sizeof(uint32_t);
+
 
 #if DMTR_DEBUG
     printf("recv: sga_numsegs: %d\n", sga.sga_numsegs);
 #endif
 
     if (sga.sga_numsegs != 0xdeadbeef) {
+        // Allocate enough space for all sga segments
+        size_t sga_total_size = ntohs(ip_hdr->total_length) - sizeof(*ip_hdr) - sizeof(*udp_hdr) - sizeof(uint32_t);
+        DMTR_OK(dmtr_malloc(&sga.sga_buf, sga_total_size));
+
+        // Read into the sga buffer
+        const void *read_buf = rte_pktmbuf_read(pkt, offset, sga_total_size, sga.sga_buf);
+        // If read_buf is not pointing to sga_buf, it's because no copy was necessary
+        // (to avoid freeing problems later, copy into sga_buf anyway)
+        if (read_buf != sga.sga_buf) {
+            rte_memcpy(sga.sga_buf, read_buf, sga_total_size);
```

Comment:
Can't we just free sga.sga_buf and set it to be read_buf? The documentation says when no copy is necessary, it's because the data was continuous (in one segment, I assume). We would have to, instead of free(), call the mbuf free to return the mbuf to the pool, but that can be part of a dmtr_free() call that knows how to distinguish a mbuf from a hoard/glibc malloc'ed returned pointer. A mbuf would not be freed, as tx_burst would free it automatically.

That's is likely not beneficial if freeing is more expensive than copying (which I assume not w/o really being sure)